sandbox/host/src/sandbox/wasmi_backend.rs

```diff
@@ -18,393 +18,544 @@
-        // should happen through the wrapper, that enforces the memory access protocol.
-        let mem = wrapper.0;
-
-        Ok(mem)
+impl Default for Backend {
+    fn default() -> Self {
+        Self::new()
     }
+}
 
-    fn resolve_global(
-        &self,
-        module_name: &str,
-        field_name: &str,
-        _global_type: &sandbox_wasmi::GlobalDescriptor,
-    ) -> std::result::Result<sandbox_wasmi::GlobalRef, sandbox_wasmi::Error> {
-        Err(sandbox_wasmi::Error::Instantiation(format!(
-            "Export {}:{} not found",
-            module_name, field_name
-        )))
+impl Backend {
+    pub fn new() -> Self {
+        let register_len = size_of::<UntypedVal>();
+
+        const DEFAULT_MIN_VALUE_STACK_HEIGHT: usize = 1024;
+        const DEFAULT_MAX_VALUE_STACK_HEIGHT: usize = 1024 * DEFAULT_MIN_VALUE_STACK_HEIGHT;
+        const DEFAULT_MAX_RECURSION_DEPTH: usize = 16384;
+
+        let mut config = Config::default();
+        config.set_stack_limits(
+            StackLimits::new(
+                DEFAULT_MIN_VALUE_STACK_HEIGHT / register_len,
+                DEFAULT_MAX_VALUE_STACK_HEIGHT / register_len,
+                DEFAULT_MAX_RECURSION_DEPTH,
+            )
+            .expect("infallible"),
+        );
+
+        let engine = Engine::new(&config);
+        let store = Store::new(&engine, None);
+        Backend {
+            store: Rc::new(StoreRefCell::new(store)),
+        }
     }
 
-    fn resolve_table(
-        &self,
-        module_name: &str,
-        field_name: &str,
-        _table_type: &sandbox_wasmi::TableDescriptor,
-    ) -> std::result::Result<sandbox_wasmi::TableRef, sandbox_wasmi::Error> {
-        Err(sandbox_wasmi::Error::Instantiation(format!(
-            "Export {}:{} not found",
-            module_name, field_name
-        )))
+    pub fn store(&self) -> &Rc<StoreRefCell> {
+        &self.store
     }
 }
 
 /// Allocate new memory region
-pub fn new_memory(initial: u32, maximum: Option<u32>) -> crate::error::Result<Memory> {
-    let memory = Memory::Wasmi(MemoryWrapper::new(
-        MemoryInstance::alloc(Pages(initial as usize), maximum.map(|m| Pages(m as usize)))
-            .map_err(|error| Error::Sandbox(error.to_string()))?,
-    ));
-
-    Ok(memory)
+pub fn new_memory(
+    store: Rc<StoreRefCell>,
+    initial: u32,
+    maximum: Option<u32>,
+) -> crate::error::Result<(Memory, Allocation)> {
+    let ty =
+        MemoryType::new(initial, maximum).map_err(|error| Error::Sandbox(error.to_string()))?;
+    let mut alloc = region::alloc(u32::MAX as usize, Protection::READ_WRITE)
+        .unwrap_or_else(|err| unreachable!("Failed to allocate memory: {err}"));
+    // # Safety:
+    //
+    // `wasmi::Memory::new_static()` requires static lifetime so we convert our buffer to it
+    // but actual lifetime of the buffer is lifetime of `Store<T>` itself,
+    // so memory will be deallocated when `Store<T>` is dropped.
```

Comment:
Update comment according to actual structures hierarchy

Reply:
Fixed